documents/dev_designs/tachyon_system.md

```diff
@@ -0,0 +1,220 @@
+  May be in memory process (potentially partitionned) or just saving to the DB.
+
+
+```mermaid
+
+sequenceDiagram
+    participant P1 as Player1
+    participant P2 as Player2
+    participant Lobby
+    participant ChatRoom
+    participant LobbyList
+
+    P1 ->> Lobby: subscribe
+    P1 ->> +Lobby: create lobby
+    Lobby ->> ChatRoom: setup chatroom
+    Lobby ->> -P1: {:ok, lobby state}
+
+    P2 ->> Lobby: subscribe
+    P2 ->> +Lobby: join lobby
+    Lobby ->> -P2: {:ok, lobby state}
+
+    P2 ->> +Lobby: rename
+    Lobby ->> -P2: :ok
+    Lobby ->> P1: broadcast update
+    Lobby ->> P2: broadcast update
+    Lobby ->> LobbyList: cast update
+
+```
+
+### Disconnection and leaving a lobby
+
+Lobbies monitor the player session process to check if they are still connected.
+When that process dies, the player should be removed from the lobby.
+
+
+
+
+## Matchmaking, player and lobby interactions
+
+Matchmaking can be done using the approach described in [the matchmaking dev guide](./matchmaking.md).
+A matchmaking queue is a process, global across the cluster, spawned at startup.
+For MVP the queue is a single process, but this could be partitionned by buckets later.
+For MVP, the queues will be hardcoded, but that could be DB driven in the future.
+
+Entities:
+* P1, P2: PlayerConnections process directly handling tachyon commands.
+* MMQueue: one of the matchmaking queue process.
+* Lobby: a lobby process. Handle player list, bosses, votes, bosses, game settings like map, ranked?, tweaks.
+* Autohost: represent a pool of connected autohost ready to host a game.
+
+```mermaid
+sequenceDiagram
+    participant P1 as Player1
+    participant P2 as Player2
+    participant MM as MMQueue
+    participant Lobby
+    participant Autohost
+
+    P1 ->> MM: join
+    P2 ->> MM: join
+    MM ->> P1: match found with P2
+    MM ->> P2: match found with P1
+    P1 ->> MM: accept
+    P2 ->> MM: accept
+    MM ->> Lobby: create with settings & players
+    Lobby ->> MM: {:ok, <lobby_id>}
+    Lobby ->> Autohost: request autohost
+    Autohost ->> Lobby: :ok
+    MM ->> P1: join lobby <lobby_id>
+    MM ->> P2: join lobby <lobby_id>
+    P1 ->> Lobby: join
+    P2 ->> Lobby: join
+
+    Lobby ->> Lobby: make start script
+    Lobby ->> Autohost: start game
+    Lobby ->> P1: start game
+    Lobby ->> P2: start game
+```
+
+If a player fail to answer a ready request within a timeout, that player is
+evicted from the queue, and all other players are put back where they were.
+
+
+## Listings
+
+We need a way to list online players + some info, as well as the lobbies. This
+should also support some kind of filtering.
+
+For example:
+* show me all public lobbies from 4v4 and up, with no OS restriction and 1 min chevron.
+* show me all my friends and their status (playing in game, spectating, chilling, in lobby)
+
+This part is the most speculative as of yet. There are basically 3 options:
```

Comment:
I agree it's speculative, we could probably start with option 2 and see to how many messages etc does this part need to scale. For option 1, it depends on database backend e.g. I'm not sure it will do better then option 2 that is purely in memory.